lib/Onyx.js

```diff
@@ -794,6 +802,27 @@ function notifyCollectionSubscribersOnNextTick(key, value) {
     Promise.resolve().then(() => keysChanged(key, value));
 }
 
+/**
+ * This method waits for prior modifications to complete before running the next modification.
+ *
+ * @param {Function} runModification
+ * @param {String} key If "key" is not set (or set to ALL_KEYS_OPERATION) then the modification will be queued
+ * until all prior modifications complete. After that, a new modification with key -1 will be queued.
+ * @returns {Promise}
+ */
+function queueModification(runModification, key = ALL_KEYS_OPERATION) {
+    const currentTask = key === ALL_KEYS_OPERATION
+        ? Promise.all(_.values(keyOperations))
+        : Promise.all([
+            keyOperations[ALL_KEYS_OPERATION],
+            keyOperations[key],
+        ]);
```

Comment:
> From past experience, we've had situations where `clear` was executed after a logout action. Some pending data merges, which were outdated, were mistakenly combined into the storage that was just cleared. This resulted in a bug. A potential issue arises if we queue write operations to occur after `clear` – it could reintroduce this bug.

This is very important. In my opinion Onyx should work in such a way that clear is absolute, it should do what it says. If I make a bunch of calls in App to Onyx.merge with some massive objects, and then the user is signed out and we clear Onyx, we don't want those pending writes to happen at all, or at least they should not happen after clearing. I think the current implementation achieves that by getting all keys before clearing.

It would be a performance optimization to cancel the pending writes and clear right away, but probably not necessary right now. I think I'm happy with these changes to push a clear operation after pending writes.

Reply:
>An earlier suggestion proposed that the clear command should abort any pending operations. However, this could be quite complex to implement, and it might be best to address this in a separate Pull Request.

I guess I'm just not sure how we got to this point or why we started worrying about the clear method and queuing things at all in this PR. How does it relate to implementing JSON_PATCH()? Are we still focused on fixing a single problem here? Can someone TL;DR?

Reply:
Added a more detailed explanation [in this comment](https://github.com/Expensify/react-native-onyx/pull/238#issuecomment-1584173684)